networkx/algorithms/approximation/tests/test_traveling_salesman.py

```diff
@@ -0,0 +1,368 @@
+
+
+class TestSimulatedAnnealingTSP:
+    setup_class = _setup_class
+
+    def test_simulated_annealing(self):
+        cycle = nx_app.simulated_annealing_tsp(self.DG, source="D")
+        cost = sum(self.DG[n][nbr]["weight"] for n, nbr in pairwise(cycle))
+        validate_solution(cycle, cost, ["D", "C", "B", "A", "D"], 31.0)
+
+        cycle = nx_app.simulated_annealing_tsp(self.DG2, source="D")
+        cost = sum(self.DG2[n][nbr]["weight"] for n, nbr in pairwise(cycle))
+        validate_solution(cycle, cost, ["D", "A", "B", "C", "D"], 53.0)
+        cycle = nx_app.simulated_annealing_tsp(self.DG2, move="1-0", source="D")
+        cost = sum(self.DG2[n][nbr]["weight"] for n, nbr in pairwise(cycle))
+        validate_solution(cycle, cost, ["D", "A", "B", "C", "D"], 53.0)
+        cycle = nx_app.simulated_annealing_tsp(self.UG2, source="D")
+        cost = sum(self.UG2[n][nbr]["weight"] for n, nbr in pairwise(cycle))
+        validate_symmetric_solution(cycle, cost, ["D", "C", "B", "A", "D"], 25.0)
+        cycle = nx_app.simulated_annealing_tsp(self.UG2, move="1-0", source="D")
+        cost = sum(self.UG2[n][nbr]["weight"] for n, nbr in pairwise(cycle))
+        validate_symmetric_solution(cycle, cost, ["D", "C", "B", "A", "D"], 25.0)
+
+        initial_sol = ["D", "B", "A", "C", "D"]
+        cycle = nx_app.simulated_annealing_tsp(self.DG, cycle=initial_sol, source="D")
+        cost = sum(self.DG[n][nbr]["weight"] for n, nbr in pairwise(cycle))
+        validate_solution(cycle, cost, ["D", "C", "B", "A", "D"], 31.0)
+        initial_sol = ["D", "A", "C", "B", "D"]
+        cycle = nx_app.simulated_annealing_tsp(
+            self.DG, move="1-0", cycle=initial_sol, source="D"
+        )
+        cost = sum(self.DG[n][nbr]["weight"] for n, nbr in pairwise(cycle))
+        validate_solution(cycle, cost, ["D", "C", "B", "A", "D"], 31.0)
+
+        cycle = nx_app.simulated_annealing_tsp(self.UG, source="D")
+        cost = sum(self.UG[n][nbr]["weight"] for n, nbr in pairwise(cycle))
+        validate_solution(cycle, cost, ["D", "C", "B", "A", "D"], 33.0)
+
+    def test_not_complete_graph(self):
+        pytest.raises(
+            nx.NetworkXError,
+            nx_app.simulated_annealing_tsp,
+            self.incompleteUG,
+            source=0,
+            cycle=[],
+        )
+        pytest.raises(
+            nx.NetworkXError,
+            nx_app.simulated_annealing_tsp,
+            self.incompleteDG,
+            source=0,
+            cycle=[],
+        )
+
+    def test_not_weighted_graph(self):
+        pytest.raises(
+            nx.NetworkXError,
+            nx_app.simulated_annealing_tsp,
+            self.unweightedUG,
+            source=0,
+            cycle=[],
+        )
+        pytest.raises(
+            nx.NetworkXError,
+            nx_app.simulated_annealing_tsp,
+            self.unweightedDG,
+            source=0,
+            cycle=[],
+        )
+
+    def test_simulated_annealing_fails(self):
```

Comment:
It might be good to add a docstring to this test explaining what the expected failure mode is (too few iterations, bad tolerance, etc). If it's a concise reason, giving the test a more descriptive name would work too, e.g. `def test_simulated_annealing_fails_too_few_iterations(self)`